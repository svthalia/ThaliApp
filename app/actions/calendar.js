export const OPEN = 'CALENDAR_OPEN';
export const REFRESH = 'CALENDAR_REFRESH';
export const SUCCESS = 'CALENDAR_SUCCESS';
export const FAILURE = 'CALENDAR_FAILURE';
export const FETCHING = 'CALENDAR_FETCHING';

export function open() {
  return {
    type: OPEN,
  };
}

export function refresh(keywords) {
  return {
    type: REFRESH,
    payload: { keywords },
  };
}

export function success(eventList) {
  return {
    type: SUCCESS,
    payload: { eventList },
  };
}

export function failure() {
  return { type: FAILURE };
}

export function fetching() {
  return {
    type: FETCHING,
  };
}
