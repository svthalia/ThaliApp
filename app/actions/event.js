export const EVENT = 'EVENT_EVENT';
export const OPEN = 'EVENT_OPEN';
export const FETCHING = 'EVENT_FETCHING';
export const SUCCESS = 'EVENT_SUCCESS';
export const DONE = 'EVENT_DONE';
export const FAILURE = 'EVENT_FAILURE';
export const ADMIN = 'EVENT_ADMIN';
export const UPDATE_REGISTRATION = 'EVENT_UPDATE_REGISTRATION';
export const ADD_TO_CALENDAR = 'EVENT_ADD_CALENDAR';

export function event(pk, navigateToEventScreen = true) {
  return {
    type: EVENT,
    payload: { pk, navigateToEventScreen },
  };
}

export function open() {
  return {
    type: OPEN,
  };
}

export function success(eventData, eventRegistrations) {
  return {
    type: SUCCESS,
    payload: { eventData, eventRegistrations },
  };
}

export function done() {
  return {
    type: DONE,
  };
}

export function fetching() {
  return {
    type: FETCHING,
  };
}

export function failure() {
  return {
    type: FAILURE,
  };
}

export function admin() {
  return {
    type: ADMIN,
  };
}

export function updateRegistration(pk, present, payment) {
  return {
    type: UPDATE_REGISTRATION,
    payload: { pk, present, payment },
  };
}

export function addToCalendar(eventName, eventLocation, eventStart, eventEnd) {
  return {
    type: ADD_TO_CALENDAR,
    payload: {
      eventName,
      eventLocation,
      eventStart,
      eventEnd,
    },
  };
}
