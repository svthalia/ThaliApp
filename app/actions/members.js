export const OPEN = 'MEMBERS_OPEN';
export const MEMBERS = 'MEMBERS_MEMBERS';
export const FETCHING = 'MEMBERS_FETCHING';
export const MEMBERS_SUCCESS = 'MEMBERS_MEMBERS_SUCCESS';
export const FAILURE = 'MEMBERS_FAILURE';
export const MORE = 'MEMBERS_MORE';
export const MORE_SUCCESS = 'MEMBERS_MORE_SUCCESS';

export function open() {
  return {
    type: OPEN,
  };
}

export function members(keywords = '') {
  return {
    type: MEMBERS,
    payload: { keywords },
  };
}

export function fetching() {
  return {
    type: FETCHING,
  };
}

export function success(memberList, next, searchKey) {
  return {
    type: MEMBERS_SUCCESS,
    payload: { memberList, next, searchKey },
  };
}

export function failure() {
  return {
    type: FAILURE,
  };
}

export function more(url) {
  return {
    type: MORE,
    payload: { url },
  };
}

export function moreSuccess(memberList, next) {
  return {
    type: MORE_SUCCESS,
    payload: { memberList, next },
  };
}
