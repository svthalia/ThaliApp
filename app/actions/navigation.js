export const BACK = 'NAVIGATE_BACK';
export const TOGGLE_DRAWER = 'NAVIGATE_TOGGLE_DRAWER';
export const OPEN_WEBSITE = 'NAVIGATE_OPEN_WEBSITE';

export function goBack() {
  return {
    type: BACK,
  };
}

export function toggleDrawer() {
  return {
    type: TOGGLE_DRAWER,
  };
}

export function openWebsite(url) {
  return {
    type: OPEN_WEBSITE,
    payload: url,
  };
}
