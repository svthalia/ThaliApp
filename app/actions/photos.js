export const PHOTOS_ALBUMS_OPEN = 'PHOTOS_ALBUMS_OPEN';
export const PHOTOS_ALBUMS_LOAD = 'PHOTOS_ALBUMS_LOAD';
export const PHOTOS_ALBUMS_SUCCESS = 'PHOTOS_ALBUMS_SUCCESS';
export const PHOTOS_ALBUMS_FAILURE = 'PHOTOS_ALBUMS_FAILURE';
export const PHOTOS_ALBUMS_FETCHING = 'PHOTOS_ALBUMS_FETCHING';
export const PHOTOS_ALBUM_OPEN = 'PHOTOS_ALBUM_OPEN';
export const PHOTOS_ALBUM_SUCCESS = 'PHOTOS_ALBUM_SUCCESS';
export const PHOTOS_ALBUM_FAILURE = 'PHOTOS_ALBUM_FAILURE';
export const PHOTOS_ALBUM_FETCHING = 'PHOTOS_ALBUM_FETCHING';
export const PHOTOS_GALLERY_OPEN = 'PHOTOS_GALLERY_OPEN';

export function openAlbums() {
  return {
    type: PHOTOS_ALBUMS_OPEN,
  };
}

export function loadAlbums(keywords, next) {
  return {
    type: PHOTOS_ALBUMS_LOAD,
    payload: { keywords, next },
  };
}

export function successAlbums(data, next, isNext) {
  return {
    type: PHOTOS_ALBUMS_SUCCESS,
    payload: { data, next, isNext },
  };
}

export function failureAlbums() {
  return { type: PHOTOS_ALBUMS_FAILURE };
}

export function fetchingAlbums() {
  return {
    type: PHOTOS_ALBUMS_FETCHING,
  };
}

export function openGallery(selection) {
  return {
    type: PHOTOS_GALLERY_OPEN,
    payload: { selection },
  };
}

export function openAlbum(pk) {
  return {
    type: PHOTOS_ALBUM_OPEN,
    payload: { pk },
  };
}

export function openAlbumWithSlug(slug) {
  return {
    type: PHOTOS_ALBUM_OPEN,
    payload: { slug },
  };
}

export function successAlbum(data) {
  return {
    type: PHOTOS_ALBUM_SUCCESS,
    payload: { data },
  };
}

export function failureAlbum() {
  return { type: PHOTOS_ALBUM_FAILURE };
}

export function fetchingAlbum() {
  return {
    type: PHOTOS_ALBUM_FETCHING,
  };
}
