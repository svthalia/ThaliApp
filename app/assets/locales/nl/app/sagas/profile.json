{
  "Change profile picture": "Profielafbeelding veranderen",
  "Uploading your new profile picture...": "Nieuwe profielafbeelding uploaden...",
  "Could not update profile picture": "De profielafbeelding kon niet worden bijgewerkt"
}
