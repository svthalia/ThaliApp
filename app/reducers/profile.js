import * as profileActions from '../actions/profile';

import { DEFAULT_PROFILE_PHOTO } from '../constants';

const initialState = {
  profile: {
    pk: -1,
    display_name: '',
    avatar: {
      full: DEFAULT_PROFILE_PHOTO,
      large: DEFAULT_PROFILE_PHOTO,
      medium: DEFAULT_PROFILE_PHOTO,
      small: DEFAULT_PROFILE_PHOTO,
    },
    profile_description: '',
    birthday: '',
    starting_year: -1,
    programme: '',
    website: '',
    membership_type: '',
    achievements: [],
    societies: [],
  },
  success: false,
  hasLoaded: false,
};

export default function profile(state = initialState, { type, payload } = {}) {
  switch (type) {
    case profileActions.FETCHING:
      return {
        ...state,
        hasLoaded: false,
      };
    case profileActions.SUCCESS:
      return {
        ...state,
        profile: {
          ...state.profile,
          ...payload,
        },
        success: true,
        hasLoaded: true,
      };
    case profileActions.FAILURE:
      return {
        ...state,
        success: false,
        hasLoaded: true,
      };
    case profileActions.UPDATE_SUCCESS:
      return {
        ...state,
        profile: {
          ...state.profile,
          ...payload,
        },
      };
    default:
      return state;
  }
}
