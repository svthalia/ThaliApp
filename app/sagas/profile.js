import ImagePicker from 'react-native-image-crop-picker';
import {
  call, put, select, takeEvery,
} from 'redux-saga/effects';
import { Alert } from 'react-native';
import Snackbar from 'react-native-snackbar';
import i18next from '../utils/i18n';

import { apiRequest } from '../utils/url';
import * as profileActions from '../actions/profile';
import { tokenSelector } from '../selectors/session';
import reportError from '../utils/errorReporting';
import * as sessionActions from '../actions/session';

const t = i18next.getFixedT(undefined, 'sagas/profile');

function* profile(action) {
  const { member } = action.payload;
  const token = yield select(tokenSelector);

  yield put(profileActions.fetching());

  const data = {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      Authorization: `Token ${token}`,
    },
  };

  try {
    const profileData = yield call(apiRequest, `members/${member}`, data);
    yield put(profileActions.success(profileData));
  } catch (error) {
    yield call(reportError, error);
    yield put(profileActions.failure());
  }
}

function* updateAvatar() {
  const options = {
    title: t('Change profile picture'),
    width: 500,
    height: 500,
    cropping: true,
    mediaType: 'photo',
  };

  try {
    const photo = yield call([ImagePicker, 'openPicker'], options);
    const fileName = photo.path.split('/').pop();

    const formData = new FormData();

    formData.append('photo', {
      name: fileName,
      type: photo.mime,
      uri: photo.path,
    });

    const token = yield select(tokenSelector);

    const data = {
      method: 'PATCH',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'multipart/form-data',
        Authorization: `Token ${token}`,
      },
      body: formData,
    };

    try {
      yield call([Snackbar, 'show'], { text: t('Uploading your new profile picture...'), duration: Snackbar.LENGTH_INDEFINITE });
      const profileData = yield call(apiRequest, 'members/me', data);
      yield call([Snackbar, 'dismiss']);
      yield put(profileActions.success(profileData));
      yield put(sessionActions.fetchUserInfo());
    } catch (error) {
      yield call([Snackbar, 'dismiss']);
      yield call(reportError, error);
      if ('photo' in error.response.jsonData) {
        yield call(Alert.alert, t('Could not update profile picture'), error.response.jsonData.photo.join(' '));
      } else {
        yield call([Snackbar, 'show'], { text: t('Could not update profile picture') });
      }
    }
  } catch (e) {
    // eat error, om nom nom
    // error from the picker that we cannot do anything about
  }
}

export default function* () {
  yield takeEvery(profileActions.PROFILE, profile);
  yield takeEvery(profileActions.CHANGE_AVATAR, updateAvatar);
}
