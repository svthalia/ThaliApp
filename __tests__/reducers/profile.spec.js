import reducer from '../../app/reducers/profile';
import * as actions from '../../app/actions/profile';

describe('profile reducer', () => {
  const emptyState = {};

  describe('initially', () => {
    const initialState = reducer();

    it('should return the initial state', () => {
      expect(initialState).toMatchSnapshot();
    });
  });

  describe('is fetching', () => {
    const state = reducer(emptyState, actions.fetching());

    it('should be loading', () => {
      expect(state).toHaveProperty('hasLoaded', false);
    });
  });

  describe('is successful', () => {
    const state = reducer(emptyState, actions.success({ pk: 1 }));

    it('should not be loading', () => {
      expect(state).toHaveProperty('hasLoaded', true);
    });

    it('should be successful', () => {
      expect(state).toHaveProperty('success', true);
    });

    it('should contain the profile data', () => {
      expect(state).toHaveProperty('profile', { pk: 1 });
    });
  });

  describe('is failure', () => {
    const state = reducer(emptyState, actions.failure());

    it('should not be loading', () => {
      expect(state).toHaveProperty('hasLoaded', true);
    });

    it('should not be successful', () => {
      expect(state).toHaveProperty('success', false);
    });
  });
});
