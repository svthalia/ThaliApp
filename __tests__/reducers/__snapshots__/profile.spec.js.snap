// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profile reducer initially should return the initial state 1`] = `
Object {
  "hasLoaded": false,
  "profile": Object {
    "achievements": Array [],
    "avatar": Object {
      "full": "http://localhost:8000/static/members/images/default-avatar.jpg",
      "large": "http://localhost:8000/static/members/images/default-avatar.jpg",
      "medium": "http://localhost:8000/static/members/images/default-avatar.jpg",
      "small": "http://localhost:8000/static/members/images/default-avatar.jpg",
    },
    "birthday": "",
    "display_name": "",
    "membership_type": "",
    "pk": -1,
    "profile_description": "",
    "programme": "",
    "societies": Array [],
    "starting_year": -1,
    "website": "",
  },
  "success": false,
}
`;
